service/round_test.go

```diff
@@ -2,227 +2,322 @@ package service
+		),
+	)
+}
 
-	// Execute r2, and request shutdown before completion.
-	r2, err := newRound(tmpdir, 1)
-	req.NoError(err)
-	req.NoError(r2.open())
-	req.Equal(0, r2.numChallenges())
-	req.True(r2.isEmpty())
+func TestRound_TearDown(t *testing.T) {
+	t.Parallel()
+	req := require.New(t)
 
-	for _, ch := range challenges {
-		req.NoError(r2.submit(ch, ch))
-	}
-	req.Equal(len(challenges), r2.numChallenges())
-	req.False(r2.isEmpty())
+	t.Run("no cleanup", func(t *testing.T) {
+		t.Parallel()
+		// Arrange
+		round, err := newRound(t.TempDir(), 0)
+		req.NoError(err)
+
+		// Act
+		req.NoError(round.teardown(false))
+
+		// Verify
+		_, err = os.Stat(round.datadir)
+		req.NoError(err)
+	})
+
+	t.Run("cleanup", func(t *testing.T) {
+		t.Parallel()
+		// Arrange
+		round, err := newRound(t.TempDir(), 0)
+		req.NoError(err)
+
+		// Act
+		req.NoError(round.teardown(true))
+
+		// Verify
+		_, err = os.Stat(round.datadir)
+		req.ErrorIs(err, os.ErrNotExist)
+	})
+}
 
-	stop()
-	req.ErrorIs(r2.execute(ctx, time.Now().Add(duration), prover.LowestMerkleMinMemoryLayer), prover.ErrShutdownRequested)
-	req.NoError(r2.teardown(false))
+// Test creating new round.
+func TestRound_New(t *testing.T) {
+	t.Parallel()
+	req := require.New(t)
 
-	// Recover r2 execution, and request shutdown before completion.
-	ctx, stop = context.WithCancel(context.Background())
-	defer stop()
-	r2recovery1, err := newRound(tmpdir, 1)
+	// Act
+	round, err := newRound(t.TempDir(), 7)
 	req.NoError(err)
-	req.Equal(len(challenges), r2recovery1.numChallenges())
-	req.False(r2recovery1.isEmpty())
+	t.Cleanup(func() {
+		req.NoError(round.teardown(true))
+	})
+
+	// Verify
+	req.EqualValues(7, round.Epoch())
+	req.True(round.isOpen())
+	req.False(round.isExecuted())
+	req.Zero(round.roundState.ExecutionStarted)
+	req.Zero(numChallenges(round))
+}
 
-	state, err := r2recovery1.state()
-	req.NoError(err)
+// Test submitting many challenges.
+func TestRound_Submit(t *testing.T) {
+	t.Parallel()
+	req := require.New(t)
 
-	stop()
-	req.ErrorIs(r2recovery1.recoverExecution(ctx, state.Execution, time.Now().Add(duration)), prover.ErrShutdownRequested)
-	req.NoError(r2recovery1.teardown(false))
+	t.Run("submit many different challenges", func(t *testing.T) {
+		t.Parallel()
+		// Arrange
+		round := newTestRound(t)
+		challenges, err := genChallenges(32)
+		req.NoError(err)
 
-	// Recover r2 execution again, and let it complete.
-	ctx, stop = context.WithCancel(context.Background())
-	defer stop()
-	r2recovery2, err := newRound(tmpdir, 1)
-	req.NoError(err)
-	req.Equal(len(challenges), r2recovery2.numChallenges())
-	req.False(r2recovery2.isEmpty())
-	state, err = r2recovery2.state()
-	req.NoError(err)
+		// Act
+		for _, ch := range challenges {
+			req.NoError(round.submit(ch, ch))
+		}
 
-	req.NoError(r2recovery2.recoverExecution(ctx, state.Execution, time.Now().Add(duration)))
-	req.NoError(r2recovery2.teardown(true))
+		// Verify
+		req.EqualValues(len(challenges), numChallenges(round))
```

Comment:
I think both are `int` and should also pass `req.Equal()`.

This should be the case for most uses of `EqualValues` in this file.